diff --git a/sdk/android/api/org/webrtc/RtpParameters.java b/sdk/android/api/org/webrtc/RtpParameters.java
index d6a22662bb..a0bfdc65a5 100644
--- a/sdk/android/api/org/webrtc/RtpParameters.java
+++ b/sdk/android/api/org/webrtc/RtpParameters.java
@@ -43,6 +43,27 @@ public class RtpParameters {
     }
   }
 
+  public static class ResolutionRestriction {
+    public int maxWidth;
+    public int maxHeight;
+
+    @CalledByNative("ResolutionRestriction")
+    ResolutionRestriction(int maxWidth, int maxHeight) {
+      this.maxWidth = maxWidth;
+      this.maxHeight = maxHeight;
+    }
+
+    @CalledByNative("ResolutionRestriction")
+    int getMaxWidth() {
+      return maxWidth;
+    }
+
+    @CalledByNative("ResolutionRestriction")
+    int getMaxHeight() {
+      return maxHeight;
+    }
+  }
+
   public static class Encoding {
     // If non-null, this represents the RID that identifies this encoding layer.
     // RIDs are used to identify layers in simulcast.
@@ -76,6 +97,7 @@ public class RtpParameters {
     // If non-null, scale the width and height down by this factor for video. If null,
     // implementation default scaling factor will be used.
     @Nullable public Double scaleResolutionDownBy;
+    @Nullable public ResolutionRestriction scaleResolutionDownTo;
     @Nullable public String scalabilityMode;
     // SSRC to be used by this encoding.
     // Can't be changed between getParameters/setParameters.
@@ -94,7 +116,7 @@ public class RtpParameters {
     @CalledByNative("Encoding")
     Encoding(String rid, boolean active, double bitratePriority, @Priority int networkPriority,
         Integer maxBitrateBps, Integer minBitrateBps, Integer maxFramerate,
-        Integer numTemporalLayers, Double scaleResolutionDownBy, String scalabilityMode, Long ssrc,
+        Integer numTemporalLayers, Double scaleResolutionDownBy, ResolutionRestriction scaleResolutionDownTo, String scalabilityMode, Long ssrc,
         boolean adaptiveAudioPacketTime) {
       this.rid = rid;
       this.active = active;
@@ -105,6 +127,7 @@ public class RtpParameters {
       this.maxFramerate = maxFramerate;
       this.numTemporalLayers = numTemporalLayers;
       this.scaleResolutionDownBy = scaleResolutionDownBy;
+      this.scaleResolutionDownTo = scaleResolutionDownTo;
       this.scalabilityMode = scalabilityMode;
       this.ssrc = ssrc;
       this.adaptiveAudioPacketTime = adaptiveAudioPacketTime;
@@ -168,6 +191,12 @@ public class RtpParameters {
       return scaleResolutionDownBy;
     }
 
+    @Nullable
+    @CalledByNative("Encoding")
+    ResolutionRestriction getScaleResolutionDownTo() {
+      return scaleResolutionDownTo;
+    }
+
     @CalledByNative("Encoding")
     Long getSsrc() {
       return ssrc;
diff --git a/sdk/android/src/jni/pc/rtp_parameters.cc b/sdk/android/src/jni/pc/rtp_parameters.cc
index dfc7f17f6d..a7858b0dcf 100644
--- a/sdk/android/src/jni/pc/rtp_parameters.cc
+++ b/sdk/android/src/jni/pc/rtp_parameters.cc
@@ -42,6 +42,24 @@ webrtc::DegradationPreference JavaToNativeDegradationPreference(
   return webrtc::DegradationPreference::DISABLED;
 }
 
+ScopedJavaLocalRef<jobject> NativeToJavaResolutionRestriction(
+    JNIEnv* env,
+    const webrtc::Resolution& resolution) {
+  return Java_ResolutionRestriction_Constructor(
+      env, resolution.width, resolution.height);
+}
+
+webrtc::Resolution JavaToNativeResolutionRestriction(
+    JNIEnv* jni,
+    const JavaRef<jobject>& j_resolution) {
+  jint width = Java_ResolutionRestriction_getMaxWidth(jni, j_resolution);
+  jint height = Java_ResolutionRestriction_getMaxHeight(jni, j_resolution);
+  webrtc::Resolution resolution;
+  resolution.width = width;
+  resolution.height = height;
+  return resolution;
+}
+
 ScopedJavaLocalRef<jobject> NativeToJavaRtpEncodingParameter(
     JNIEnv* env,
     const RtpEncodingParameters& encoding) {
@@ -53,6 +71,7 @@ ScopedJavaLocalRef<jobject> NativeToJavaRtpEncodingParameter(
       NativeToJavaInteger(env, encoding.max_framerate),
       NativeToJavaInteger(env, encoding.num_temporal_layers),
       NativeToJavaDouble(env, encoding.scale_resolution_down_by),
+      encoding.scale_resolution_down_to ? NativeToJavaResolutionRestriction(env, *encoding.scale_resolution_down_to) : nullptr,
       NativeToJavaString(env, encoding.scalability_mode),
       encoding.ssrc ? NativeToJavaLong(env, *encoding.ssrc) : nullptr,
       encoding.adaptive_ptime);
@@ -117,6 +136,11 @@ RtpEncodingParameters JavaToNativeRtpEncodingParameters(
       Java_Encoding_getScaleResolutionDownBy(jni, j_encoding_parameters);
   encoding.scale_resolution_down_by =
       JavaToNativeOptionalDouble(jni, j_scale_resolution_down_by);
+  ScopedJavaLocalRef<jobject> j_scale_resolution_down_to =
+      Java_Encoding_getScaleResolutionDownTo(jni, j_encoding_parameters);
+  if (!IsNull(jni, j_scale_resolution_down_to)) {
+    encoding.scale_resolution_down_to = JavaToNativeResolutionRestriction(jni, j_scale_resolution_down_to);
+  }
   ScopedJavaLocalRef<jstring> j_scalability_mode =
       Java_Encoding_getScalabilityMode(jni, j_encoding_parameters);
   if (!IsNull(jni, j_scalability_mode)) {
diff --git a/sdk/objc/api/peerconnection/RTCRtpEncodingParameters.h b/sdk/objc/api/peerconnection/RTCRtpEncodingParameters.h
index 467b51de98..5a2ce40afe 100644
--- a/sdk/objc/api/peerconnection/RTCRtpEncodingParameters.h
+++ b/sdk/objc/api/peerconnection/RTCRtpEncodingParameters.h
@@ -22,6 +22,12 @@ typedef NS_ENUM(NSInteger, RTCPriority) {
   RTCPriorityHigh
 };
 
+RTC_OBJC_EXPORT
+@interface RTC_OBJC_TYPE (RTCResolutionRestriction) : NSObject
+@property(nonatomic, assign) NSNumber* maxWidth;
+@property(nonatomic, assign) NSNumber* maxHeight;
+@end
+
 RTC_OBJC_EXPORT
 @interface RTC_OBJC_TYPE (RTCRtpEncodingParameters) : NSObject
 
@@ -56,6 +62,8 @@ RTC_OBJC_EXPORT
  */
 @property(nonatomic, copy, nullable) NSNumber *scaleResolutionDownBy;
 
+@property(nonatomic, copy, nullable) RTCResolutionRestriction *scaleResolutionDownTo;
+
 /** The SSRC being used by this encoding. */
 @property(nonatomic, readonly, nullable) NSNumber *ssrc;
 
diff --git a/sdk/objc/api/peerconnection/RTCRtpEncodingParameters.mm b/sdk/objc/api/peerconnection/RTCRtpEncodingParameters.mm
index 69f8885f4c..6ddea3b85a 100644
--- a/sdk/objc/api/peerconnection/RTCRtpEncodingParameters.mm
+++ b/sdk/objc/api/peerconnection/RTCRtpEncodingParameters.mm
@@ -12,6 +12,13 @@
 
 #import "helpers/NSString+StdString.h"
 
+@implementation RTC_OBJC_TYPE (RTCResolutionRestriction)
+
+@synthesize maxWidth = _maxWidth;
+@synthesize maxHeight = _maxHeight;
+
+@end
+
 @implementation RTC_OBJC_TYPE (RTCRtpEncodingParameters)
 
 @synthesize rid = _rid;
@@ -21,6 +28,7 @@
 @synthesize maxFramerate = _maxFramerate;
 @synthesize numTemporalLayers = _numTemporalLayers;
 @synthesize scaleResolutionDownBy = _scaleResolutionDownBy;
+@synthesize scaleResolutionDownTo = _scaleResolutionDownTo;
 @synthesize ssrc = _ssrc;
 @synthesize bitratePriority = _bitratePriority;
 @synthesize networkPriority = _networkPriority;
@@ -57,6 +65,11 @@
       _scaleResolutionDownBy =
           [NSNumber numberWithDouble:*nativeParameters.scale_resolution_down_by];
     }
+    if (nativeParameters.scale_resolution_down_to) {
+      _scaleResolutionDownTo = [[RTCResolutionRestriction alloc] init];
+      _scaleResolutionDownTo.maxWidth = [NSNumber numberWithInt:nativeParameters.scale_resolution_down_to->width];
+      _scaleResolutionDownTo.maxHeight = [NSNumber numberWithInt:nativeParameters.scale_resolution_down_to->height];
+    }
     if (nativeParameters.ssrc) {
       _ssrc = [NSNumber numberWithUnsignedLong:*nativeParameters.ssrc];
     }
@@ -89,6 +102,11 @@
   if (_scaleResolutionDownBy != nil) {
     parameters.scale_resolution_down_by = std::optional<double>(_scaleResolutionDownBy.doubleValue);
   }
+  if (_scaleResolutionDownTo != nil) {
+    webrtc::Resolution& r = parameters.scale_resolution_down_to.emplace();
+    r.width = _scaleResolutionDownTo.maxWidth.intValue;
+    r.height = _scaleResolutionDownTo.maxHeight.intValue;
+  }
   if (_ssrc != nil) {
     parameters.ssrc = std::optional<uint32_t>(_ssrc.unsignedLongValue);
   }
