diff --git a/tools_webrtc/libs/generate_licenses.py b/tools_webrtc/libs/generate_licenses.py
index 86b4cd01f0..cf927670ee 100755
--- a/tools_webrtc/libs/generate_licenses.py
+++ b/tools_webrtc/libs/generate_licenses.py
@@ -91,6 +91,8 @@ LIB_TO_LICENSES_DICT = {
     'yasm': [],
     'ow2_asm': [],
     'jdk': [],
+
+    'dav1d': ['third_party/dav1d/LICENSE'],
 }
 
 # Third_party library _regex_ to licences mapping. Keys are regular expression
