diff --git a/DEPS b/DEPS
index 7bf91bfc8f..1007aa2384 100644
--- a/DEPS
+++ b/DEPS
@@ -2072,7 +2072,7 @@ hooks = [
   {
     'name': 'rust',
     'pattern': '.',
-    'condition': 'checkout_android or checkout_fuzzer',
+    'condition': 'checkout_fuzzer',
     'action': ['python3', 'src/tools/rust/update_rust.py'],
   },
   {
diff --git a/tools_webrtc/android/build_aar.py b/tools_webrtc/android/build_aar.py
index f1ca2e48eb..85c1411fde 100755
--- a/tools_webrtc/android/build_aar.py
+++ b/tools_webrtc/android/build_aar.py
@@ -176,8 +176,6 @@ def Build(build_dir, arch, use_remoteexec, extra_gn_args, extra_gn_switches,
         'rtc_include_tests': False,
         'target_cpu': _GetTargetCpu(arch),
         'use_remoteexec': use_remoteexec,
-        'enable_rust': True,
-        'enable_rust_cxx': True,
     }
     arm_version = _GetArmVersion(arch)
     if arm_version:
diff --git a/tools_webrtc/mb/mb_config.pyl b/tools_webrtc/mb/mb_config.pyl
index 79d372ff9f..8384d9aa32 100644
--- a/tools_webrtc/mb/mb_config.pyl
+++ b/tools_webrtc/mb/mb_config.pyl
@@ -237,75 +237,64 @@
     'android_asan_shared_release_bot_arm': [
       'android',
       'asan',
-      'chromium_base_dependency',
       'clang',
       'pure_release_bot',
       'arm',
     ],
     'android_debug_static_bot_arm': [
       'android',
-      'chromium_base_dependency',
       'debug_static_bot',
       'arm',
       'h265',
     ],
     'android_debug_static_bot_arm64': [
       'android',
-      'chromium_base_dependency',
       'debug_static_bot',
       'arm64',
       'h265',
     ],
     'android_debug_static_bot_x64': [
       'android',
-      'chromium_base_dependency',
       'debug_static_bot',
       'x64',
       'h265',
     ],
     'android_debug_static_bot_x86': [
       'android',
-      'chromium_base_dependency',
       'debug_static_bot',
       'x86',
       'h265',
     ],
     'android_pure_release_bot_arm': [
       'android',
-      'chromium_base_dependency',
       'pure_release_bot',
       'arm',
     ],
     'android_pure_release_bot_arm64': [
       'android',
-      'chromium_base_dependency',
       'pure_release_bot',
       'arm64',
     ],
     'android_release_bot_arm': [
       'android',
-      'chromium_base_dependency',
       'release_bot',
       'arm',
       'h265',
     ],
     'android_release_bot_arm64': [
       'android',
-      'chromium_base_dependency',
       'release_bot',
       'arm64',
       'h265',
     ],
     'android_release_bot_x64': [
       'android',
-      'chromium_base_dependency',
       'release_bot',
       'x64',
       'h265',
     ],
     'android_release_bot_x86': [
       'android',
-      'chromium_base_dependency',
       'release_bot',
       'x86',
       'h265',
@@ -332,7 +321,6 @@
     'disable_trace_events_android_arm': [
       'android',
       'arm',
-      'chromium_base_dependency',
       'disable_trace_events',
       'release_bot',
     ],
@@ -341,7 +329,6 @@
     'dummy_audio_file_devices_no_protobuf_android_arm': [
       'android',
       'arm',
-      'chromium_base_dependency',
       'debug_static_bot',
       'dummy_audio_file_devices',
       'no_protobuf',
@@ -424,7 +411,6 @@
     'perfetto_android_arm': [
       'android',
       'arm',
-      'chromium_base_dependency',
       'perfetto',
       'release_bot',
     ],
@@ -441,7 +427,6 @@
     'rtti_no_sctp_android_arm': [
       'android',
       'arm',
-      'chromium_base_dependency',
       'debug_static_bot',
       'no_sctp',
       'rtti',
