diff --git a/sdk/android/src/jni/video_encoder_wrapper.cc b/sdk/android/src/jni/video_encoder_wrapper.cc
index 492adb2466..dbb9c7f91c 100644
--- a/sdk/android/src/jni/video_encoder_wrapper.cc
+++ b/sdk/android/src/jni/video_encoder_wrapper.cc
@@ -55,7 +55,7 @@ namespace jni {
 
 VideoEncoderWrapper::VideoEncoderWrapper(JNIEnv* jni,
                                          const JavaRef<jobject>& j_encoder)
-    : encoder_(jni, j_encoder), int_array_class_(GetClass(jni, "[I")) {
+    : encoder_(jni, j_encoder), int_array_class_(GetClass(jni, "[I")), callback_(nullptr) {
   initialized_ = false;
   num_resets_ = 0;
 
