diff --git a/sdk/android/BUILD.gn b/sdk/android/BUILD.gn
index 0cea04f54f..e32069f7f0 100644
--- a/sdk/android/BUILD.gn
+++ b/sdk/android/BUILD.gn
@@ -40,6 +40,7 @@ if (is_android) {
       ":builtin_audio_codecs_java",
       ":camera_java",
       ":default_video_codec_factory_java",
+      ":environment_java",
       ":filevideo_java",
       ":hwcodecs_java",
       ":java_audio_device_module_java",
