diff --git a/webrtc.gni b/webrtc.gni
index 530eea6ff7..a81bda4324 100644
--- a/webrtc.gni
+++ b/webrtc.gni
@@ -97,7 +97,7 @@ declare_args() {
   # Setting this to false will require the API user to pass in their own
   # SSLCertificateVerifier to verify the certificates presented from a
   # TLS-TURN server. In return disabling this saves around 100kb in the binary.
-  rtc_builtin_ssl_root_certificates = true
+  rtc_builtin_ssl_root_certificates = false
 
   # Disable this to avoid building the Opus audio codec.
   rtc_include_opus = true
