diff --git a/rtc_base/trace_event.h b/rtc_base/trace_event.h
index 3329abd66c..2aee713841 100644
--- a/rtc_base/trace_event.h
+++ b/rtc_base/trace_event.h
@@ -20,10 +20,10 @@
 // IWYU pragma: begin_exports
 #if defined(RTC_USE_PERFETTO)
 #include "rtc_base/trace_categories.h"
-#endif
 #include "third_party/perfetto/include/perfetto/tracing/event_context.h"
 #include "third_party/perfetto/include/perfetto/tracing/track.h"
 #include "third_party/perfetto/include/perfetto/tracing/track_event_args.h"
+#endif
 // IWYU pragma: end_exports
 
 #if !defined(RTC_USE_PERFETTO)
