diff --git i/api/audio_codecs/audio_format.h w/api/audio_codecs/audio_format.h
index edccc17e7d..023ddb660d 100644
--- i/api/audio_codecs/audio_format.h
+++ w/api/audio_codecs/audio_format.h
@@ -25,7 +25,7 @@ namespace webrtc {
 
 // SDP specification for a single audio codec.
 struct RTC_EXPORT SdpAudioFormat {
-  using Parameters [[deprecated(("Use webrtc::CodecParameterMap"))]] =
+  using Parameters [[deprecated("Use webrtc::CodecParameterMap")]] =
       std::map<std::string, std::string>;
 
   SdpAudioFormat(const SdpAudioFormat&);
diff --git i/api/video_codecs/sdp_video_format.h w/api/video_codecs/sdp_video_format.h
index af9537b5a3..9ee3ed15e7 100644
--- i/api/video_codecs/sdp_video_format.h
+++ w/api/video_codecs/sdp_video_format.h
@@ -26,7 +26,7 @@ namespace webrtc {
 // SDP specification for a single video codec.
 // NOTE: This class is still under development and may change without notice.
 struct RTC_EXPORT SdpVideoFormat {
-  using Parameters [[deprecated(("Use webrtc::CodecParameterMap"))]] =
+  using Parameters [[deprecated("Use webrtc::CodecParameterMap")]] =
       std::map<std::string, std::string>;
 
   explicit SdpVideoFormat(const std::string& name);
