diff --git a/sdk/objc/components/audio/RTCAudioSession+Configuration.mm b/sdk/objc/components/audio/RTCAudioSession+Configuration.mm
index fd9054ff89..7267a79334 100644
--- a/sdk/objc/components/audio/RTCAudioSession+Configuration.mm
+++ b/sdk/objc/components/audio/RTCAudioSession+Configuration.mm
@@ -151,7 +151,8 @@ - (BOOL)setConfiguration:
     *outError = error;
   }
 
-  return error == nil;
+  //return error == nil;
+  return YES;
 }
 
 @end
diff --git a/sdk/objc/components/audio/RTCAudioSession.h b/sdk/objc/components/audio/RTCAudioSession.h
index 93b4dc2dc5..e86867a620 100644
--- a/sdk/objc/components/audio/RTCAudioSession.h
+++ b/sdk/objc/components/audio/RTCAudioSession.h
@@ -20,6 +20,7 @@ extern NSString *const kRTCAudioSessionErrorDomain;
 extern NSInteger const kRTCAudioSessionErrorLockRequired;
 /** Unknown configuration error occurred. */
 extern NSInteger const kRTCAudioSessionErrorConfiguration;
+extern NSInteger const kRTCAudioSessionErrorInputInitialization;
 
 @class RTC_OBJC_TYPE(RTCAudioSession);
 @class RTC_OBJC_TYPE(RTCAudioSessionConfiguration);
@@ -238,6 +239,8 @@ RTC_OBJC_EXPORT
 // AVAudioSession. `lockForConfiguration` must be called before using them
 // otherwise they will fail with kRTCAudioSessionErrorLockRequired.
 
+- (BOOL)setCategory:(NSString *)category
+              error:(NSError **)outError;
 - (BOOL)setCategory:(AVAudioSessionCategory)category
                mode:(AVAudioSessionMode)mode
             options:(AVAudioSessionCategoryOptions)options
@@ -263,6 +266,15 @@ RTC_OBJC_EXPORT
 - (BOOL)setOutputDataSource:(AVAudioSessionDataSourceDescription *)dataSource
                       error:(NSError **)outError;
 
+// 入力を使用する場合だけ、AudioUnit の生成前に 1 回呼ぶ。完了通知はメインキューで一度だけ行う。
+- (void)initializeInput:(nullable void (^)(NSError *_Nullable error))completionHandler;
+
+/** initialMicrophoneMute フラグの値をセットします。
+ *  initializeInput より前に指定します。
+ *  入力初期化後は NO を返します。
+ */
+- (BOOL)setInitialMicrophoneMute:(BOOL)initialMicrophoneMute;
+
 @end
 
 @interface RTC_OBJC_TYPE (RTCAudioSession)
diff --git a/sdk/objc/components/audio/RTCAudioSession.mm b/sdk/objc/components/audio/RTCAudioSession.mm
index 4c74211db0..992c2ce2d4 100644
--- a/sdk/objc/components/audio/RTCAudioSession.mm
+++ b/sdk/objc/components/audio/RTCAudioSession.mm
@@ -30,6 +30,7 @@
     @"org.webrtc.RTC_OBJC_TYPE(RTCAudioSession)";
 NSInteger const kRTCAudioSessionErrorLockRequired = -1;
 NSInteger const kRTCAudioSessionErrorConfiguration = -2;
+NSInteger const kRTCAudioSessionErrorInputInitialization = -3;
 NSString *const kRTCAudioSessionOutputVolumeSelector = @"outputVolume";
 
 namespace {
@@ -442,6 +443,11 @@ - (BOOL)setCategory:(AVAudioSessionCategory)category
                              error:outError];
 }
 
+- (BOOL)setCategory:(NSString *)category
+              error:(NSError **)outError {
+  return [self.session setCategory:category error:outError];
+}
+
 - (BOOL)setCategory:(AVAudioSessionCategory)category
         withOptions:(AVAudioSessionCategoryOptions)options
               error:(NSError **)outError {
diff --git a/sdk/objc/components/audio/RTCAudioSessionConfiguration.m b/sdk/objc/components/audio/RTCAudioSessionConfiguration.m
index c583868291..92f3697d08 100644
--- a/sdk/objc/components/audio/RTCAudioSessionConfiguration.m
+++ b/sdk/objc/components/audio/RTCAudioSessionConfiguration.m
@@ -59,7 +59,7 @@ - (instancetype)init {
     // By default, using this category implies that our app’s audio is
     // nonmixable, hence activating the session will interrupt any other
     // audio sessions which are also nonmixable.
-    _category = AVAudioSessionCategoryPlayAndRecord;
+    _category = AVAudioSessionCategoryAmbient;
 #if defined(__IPHONE_26_0) && __IPHONE_OS_VERSION_MAX_ALLOWED >= __IPHONE_26_0
     _categoryOptions = AVAudioSessionCategoryOptionAllowBluetoothHFP;
 #else
