diff --git a/sdk/BUILD.gn b/sdk/BUILD.gn
index 1a7fd8e3f6..03a7f586c3 100644
--- a/sdk/BUILD.gn
+++ b/sdk/BUILD.gn
@@ -335,9 +335,6 @@ if (is_ios || is_mac) {
         ]
 
         frameworks = [ "AudioToolbox.framework" ]
-        if (is_mac) {
-          frameworks += [ "AudioUnit.framework" ]
-        }
       }
 
       # This target exists to expose :audio_session_objc and
@@ -520,9 +517,6 @@ if (is_ios || is_mac) {
         "../rtc_base:threading",
         "../rtc_base:timeutils",
       ]
-      if (is_mac) {
-        frameworks = [ "AudioUnit.framework" ]
-      }
     }
 
     rtc_source_set("audio_device_module_error_handler") {
@@ -744,6 +738,7 @@ if (is_ios || is_mac) {
       ]
 
       deps = [
+        ":simulcast",
         ":base_native_additions_objc",
         ":base_objc",
         ":helpers_objc",
@@ -769,6 +764,22 @@ if (is_ios || is_mac) {
       }
     }
 
+    rtc_library("simulcast") {
+      sources = [
+        "objc/components/video_codec/RTCVideoEncoderFactorySimulcast.h",
+        "objc/components/video_codec/RTCVideoEncoderFactorySimulcast.mm",
+        "objc/api/video_codec/RTCVideoEncoderSimulcast.h",
+        "objc/api/video_codec/RTCVideoEncoderSimulcast.mm",
+      ]
+
+      deps = [
+        ":base_objc",
+        ":wrapped_native_codec_objc",
+        "../media:rtc_media_base",
+        "../media:rtc_simulcast_encoder_adapter",
+      ]
+    }
+
     rtc_library("vpx_codec_constants") {
       configs += [ "..:no_global_constructors" ]
       sources = [
@@ -1366,6 +1377,7 @@ if (is_ios || is_mac) {
           "objc/components/video_codec/RTCVideoDecoderH264.h",
           "objc/components/video_codec/RTCVideoEncoderFactoryH264.h",
           "objc/components/video_codec/RTCVideoEncoderH264.h",
+          "objc/components/video_codec/RTCVideoEncoderFactorySimulcast.h",
           "objc/components/video_frame_buffer/RTCCVPixelBuffer.h",
           "objc/helpers/RTCCameraPreviewView.h",
           "objc/helpers/RTCDispatcher.h",
@@ -1417,6 +1429,7 @@ if (is_ios || is_mac) {
           "objc/api/video_codec/RTCVideoEncoderVP8.h",
           "objc/api/video_codec/RTCVideoEncoderVP9.h",
           "objc/api/video_codec/RTCVideoEncoderAV1.h",
+          "objc/api/video_codec/RTCVideoEncoderSimulcast.h",
           "objc/api/video_frame_buffer/RTCNativeI420Buffer.h",
           "objc/api/video_frame_buffer/RTCNativeMutableI420Buffer.h",
         ]
@@ -1531,6 +1544,7 @@ if (is_ios || is_mac) {
           "objc/api/video_codec/RTCVideoEncoderAV1.h",
           "objc/api/video_codec/RTCVideoEncoderVP8.h",
           "objc/api/video_codec/RTCVideoEncoderVP9.h",
+          "objc/api/video_codec/RTCVideoEncoderSimulcast.h",
           "objc/api/video_frame_buffer/RTCNativeI420Buffer.h",
           "objc/api/video_frame_buffer/RTCNativeMutableI420Buffer.h",
           "objc/base/RTCCodecSpecificInfo.h",
@@ -1565,6 +1579,7 @@ if (is_ios || is_mac) {
           "objc/components/video_codec/RTCVideoDecoderH264.h",
           "objc/components/video_codec/RTCVideoEncoderFactoryH264.h",
           "objc/components/video_codec/RTCVideoEncoderH264.h",
+          "objc/components/video_codec/RTCVideoEncoderFactorySimulcast.h",
           "objc/components/video_frame_buffer/RTCCVPixelBuffer.h",
           "objc/helpers/RTCDispatcher.h",
         ]
@@ -1675,7 +1690,6 @@ if (is_ios || is_mac) {
         "../api/video:video_rtp_headers",
         "../api/video_codecs:video_codecs_api",
         "../common_video",
-        "../pc:video_track_source_proxy",
         "../rtc_base:buffer",
         "../rtc_base:logging",
         "../rtc_base:ssl",
diff --git a/sdk/objc/api/video_codec/RTCVideoEncoderSimulcast.h b/sdk/objc/api/video_codec/RTCVideoEncoderSimulcast.h
new file mode 100644
index 0000000000..4f1b55c713
--- /dev/null
+++ b/sdk/objc/api/video_codec/RTCVideoEncoderSimulcast.h
@@ -0,0 +1,13 @@
+#import "RTCMacros.h"
+#import "RTCVideoEncoder.h"
+#import "RTCVideoEncoderFactory.h"
+#import "RTCVideoCodecInfo.h"
+
+RTC_OBJC_EXPORT
+@interface RTC_OBJC_TYPE (RTCVideoEncoderSimulcast) : NSObject
+
++ (id<RTC_OBJC_TYPE(RTCVideoEncoder)>)simulcastEncoderWithPrimary:(id<RTC_OBJC_TYPE(RTCVideoEncoderFactory)>)primary
+                                                         fallback:(id<RTC_OBJC_TYPE(RTCVideoEncoderFactory)>)fallback
+                                                   videoCodecInfo:(RTC_OBJC_TYPE(RTCVideoCodecInfo) *)videoCodecInfo;
+
+@end
diff --git a/sdk/objc/api/video_codec/RTCVideoEncoderSimulcast.mm b/sdk/objc/api/video_codec/RTCVideoEncoderSimulcast.mm
new file mode 100644
index 0000000000..6f7535b729
--- /dev/null
+++ b/sdk/objc/api/video_codec/RTCVideoEncoderSimulcast.mm
@@ -0,0 +1,64 @@
+#import <Foundation/Foundation.h>
+
+#import "RTCMacros.h"
+#import "RTCVideoEncoderSimulcast.h"
+#import "RTCNativeVideoEncoder.h"
+#import "RTCNativeVideoEncoderBuilder+Native.h"
+#import "api/peerconnection/RTCVideoCodecInfo+Private.h"
+
+#include "native/api/video_encoder_factory.h"
+#include "media/engine/simulcast_encoder_adapter.h"
+
+@interface RTC_OBJC_TYPE (RTCVideoEncoderSimulcastBuilder)
+    : RTC_OBJC_TYPE(RTCNativeVideoEncoder) <RTC_OBJC_TYPE (RTCNativeVideoEncoderBuilder)> {
+
+    id<RTC_OBJC_TYPE(RTCVideoEncoderFactory)> _primary;
+    id<RTC_OBJC_TYPE(RTCVideoEncoderFactory)> _fallback;
+    RTC_OBJC_TYPE(RTCVideoCodecInfo) *_videoCodecInfo;
+}
+
+- (id)initWithPrimary:(id<RTC_OBJC_TYPE(RTCVideoEncoderFactory)>)primary
+             fallback:(id<RTC_OBJC_TYPE(RTCVideoEncoderFactory)>)fallback
+       videoCodecInfo:(RTC_OBJC_TYPE(RTCVideoCodecInfo) *)videoCodecInfo;
+
+@end
+
+@implementation RTC_OBJC_TYPE (RTCVideoEncoderSimulcastBuilder)
+
+- (std::unique_ptr<webrtc::VideoEncoder>)build:(const webrtc::Environment&)env {
+    auto nativePrimary = webrtc::ObjCToNativeVideoEncoderFactory(_primary);
+    auto nativeFallback = webrtc::ObjCToNativeVideoEncoderFactory(_fallback);
+    auto nativeFormat = [_videoCodecInfo nativeSdpVideoFormat];
+    return std::make_unique<webrtc::SimulcastEncoderAdapter>(
+            env,
+            nativePrimary.release(),
+            nativeFallback.release(),
+            std::move(nativeFormat));
+}
+
+- (id)initWithPrimary:(id<RTC_OBJC_TYPE(RTCVideoEncoderFactory)>)primary
+             fallback:(id<RTC_OBJC_TYPE(RTCVideoEncoderFactory)>)fallback
+       videoCodecInfo:(RTC_OBJC_TYPE(RTCVideoCodecInfo) *)videoCodecInfo {
+    if (self == [super init]) {
+        self->_primary = primary;
+        self->_fallback = fallback;
+        self->_videoCodecInfo = videoCodecInfo;
+    }
+    return self;
+}
+
+@end
+
+@implementation RTC_OBJC_TYPE (RTCVideoEncoderSimulcast)
+
++ (id<RTC_OBJC_TYPE(RTCVideoEncoder)>)simulcastEncoderWithPrimary:(id<RTC_OBJC_TYPE(RTCVideoEncoderFactory)>)primary
+                                                         fallback:(id<RTC_OBJC_TYPE(RTCVideoEncoderFactory)>)fallback
+                                                   videoCodecInfo:(RTC_OBJC_TYPE(RTCVideoCodecInfo) *)videoCodecInfo {
+    return [[RTC_OBJC_TYPE(RTCVideoEncoderSimulcastBuilder) alloc]
+        initWithPrimary:primary
+               fallback:fallback
+         videoCodecInfo:videoCodecInfo];
+}
+
+@end
+
diff --git a/sdk/objc/components/video_codec/RTCDefaultVideoEncoderFactory.mm b/sdk/objc/components/video_codec/RTCDefaultVideoEncoderFactory.mm
index 3cc2121804..333977668f 100644
--- a/sdk/objc/components/video_codec/RTCDefaultVideoEncoderFactory.mm
+++ b/sdk/objc/components/video_codec/RTCDefaultVideoEncoderFactory.mm
@@ -11,6 +11,7 @@
 #import "RTCDefaultVideoEncoderFactory.h"
 
 #import "RTCH264ProfileLevelId.h"
+#import "RTCVideoEncoderFactorySimulcast.h"
 #import "RTCVideoEncoderH264.h"
 #import "api/peerconnection/RTCVideoCodecInfo+Private.h"
 #import "api/video_codec/RTCVideoCodecConstants.h"
@@ -31,23 +32,7 @@ @implementation RTC_OBJC_TYPE (RTCDefaultVideoEncoderFactory)
 @synthesize preferredCodec;
 
 + (NSArray<RTC_OBJC_TYPE(RTCVideoCodecInfo) *> *)supportedCodecs {
-  NSMutableArray<RTC_OBJC_TYPE(RTCVideoCodecInfo) *> *result =
-      [NSMutableArray array];
-
-  [result
-      addObjectsFromArray:[RTC_OBJC_TYPE(RTCVideoEncoderH264) supportedCodecs]];
-  [result
-      addObjectsFromArray:[RTC_OBJC_TYPE(RTCVideoEncoderVP8) supportedCodecs]];
-
-  [result
-      addObjectsFromArray:[RTC_OBJC_TYPE(RTCVideoEncoderVP9) supportedCodecs]];
-
-#if defined(RTC_USE_LIBAOM_AV1_ENCODER)
-  [result
-      addObjectsFromArray:[RTC_OBJC_TYPE(RTCVideoEncoderAV1) supportedCodecs]];
-#endif
-
-  return result;
+  return [RTCVideoEncoderFactorySimulcast supportedCodecsForDefaultEncoderFactory];
 }
 
 - (id<RTC_OBJC_TYPE(RTCVideoEncoder)>)createEncoder:
diff --git a/sdk/objc/components/video_codec/RTCVideoEncoderFactorySimulcast.h b/sdk/objc/components/video_codec/RTCVideoEncoderFactorySimulcast.h
new file mode 100644
index 0000000000..c1c396df10
--- /dev/null
+++ b/sdk/objc/components/video_codec/RTCVideoEncoderFactorySimulcast.h
@@ -0,0 +1,18 @@
+#import <Foundation/Foundation.h>
+
+#import "RTCMacros.h"
+#import "RTCVideoEncoderFactory.h"
+
+NS_ASSUME_NONNULL_BEGIN
+
+RTC_OBJC_EXPORT
+@interface RTC_OBJC_TYPE (RTCVideoEncoderFactorySimulcast) : NSObject <RTC_OBJC_TYPE(RTCVideoEncoderFactory)>
+
+- (instancetype)initWithPrimary:(id<RTC_OBJC_TYPE(RTCVideoEncoderFactory)>)primary
+                       fallback:(id<RTC_OBJC_TYPE(RTCVideoEncoderFactory)>)fallback;
+
++ (NSArray<RTC_OBJC_TYPE(RTCVideoCodecInfo) *> *)supportedCodecsForDefaultEncoderFactory;
+
+@end
+
+NS_ASSUME_NONNULL_END
diff --git a/sdk/objc/components/video_codec/RTCVideoEncoderFactorySimulcast.mm b/sdk/objc/components/video_codec/RTCVideoEncoderFactorySimulcast.mm
new file mode 100644
index 0000000000..3f7b0f0414
--- /dev/null
+++ b/sdk/objc/components/video_codec/RTCVideoEncoderFactorySimulcast.mm
@@ -0,0 +1,116 @@
+#import <Foundation/Foundation.h>
+
+#import "RTCH264ProfileLevelId.h"
+#import "RTCMacros.h"
+#import "RTCVideoCodecInfo.h"
+#import "RTCVideoEncoderFactorySimulcast.h"
+#import "api/video_codec/RTCVideoEncoderSimulcast.h"
+#import "api/peerconnection/RTCVideoCodecInfo+Private.h"
+
+#include "absl/container/inlined_vector.h"
+#include "api/video_codecs/video_codec.h"
+#include "api/video_codecs/sdp_video_format.h"
+#include "api/video_codecs/video_codec.h"
+#include "modules/video_coding/codecs/av1/av1_svc_config.h"
+#include "modules/video_coding/codecs/h264/include/h264.h"
+#include "modules/video_coding/codecs/vp8/include/vp8.h"
+#include "modules/video_coding/codecs/vp8/vp8_scalability.h"
+#include "modules/video_coding/codecs/vp9/include/vp9.h"
+#include "media/base/media_constants.h"
+
+@interface RTC_OBJC_TYPE (RTCVideoEncoderFactorySimulcast) ()
+
+@property id<RTC_OBJC_TYPE(RTCVideoEncoderFactory)> primary;
+@property id<RTC_OBJC_TYPE(RTCVideoEncoderFactory)> fallback;
+
+@end
+
+
+@implementation RTC_OBJC_TYPE (RTCVideoEncoderFactorySimulcast)
+
+@synthesize primary = _primary;
+@synthesize fallback = _fallback;
+
+- (instancetype)initWithPrimary:(id<RTC_OBJC_TYPE(RTCVideoEncoderFactory)>)primary
+                       fallback:(id<RTC_OBJC_TYPE(RTCVideoEncoderFactory)>)fallback {
+    if (self = [super init]) {
+        _primary = primary;
+        _fallback = fallback;
+    }
+    return self;
+}
+
+- (nullable id<RTC_OBJC_TYPE(RTCVideoEncoder)>)createEncoder: (RTC_OBJC_TYPE(RTCVideoCodecInfo) *)info {
+    return [RTCVideoEncoderSimulcast simulcastEncoderWithPrimary: _primary fallback: _fallback videoCodecInfo: info];
+}
+
+- (NSArray<RTC_OBJC_TYPE(RTCVideoCodecInfo) *> *)supportedCodecs {
+    NSArray *supportedCodecs = [[_primary supportedCodecs] arrayByAddingObjectsFromArray: [_fallback supportedCodecs]];
+
+    NSMutableArray<RTC_OBJC_TYPE(RTCVideoCodecInfo) *> *addingCodecs = [[NSMutableArray alloc] init];
+
+    for (const webrtc::SdpVideoFormat& format : webrtc::SupportedVP9Codecs(true)) {
+        RTCVideoCodecInfo *codec = [[RTCVideoCodecInfo alloc] initWithNativeSdpVideoFormat: format];
+        [addingCodecs addObject: codec];
+    }
+
+    auto av1Format = webrtc::SdpVideoFormat(
+         webrtc::kAv1CodecName, webrtc::CodecParameterMap(),
+         webrtc::LibaomAv1EncoderSupportedScalabilityModes());
+    RTCVideoCodecInfo *av1Codec = [[RTCVideoCodecInfo alloc] initWithNativeSdpVideoFormat: av1Format];
+    [addingCodecs addObject: av1Codec];
+
+    return [supportedCodecs arrayByAddingObjectsFromArray: addingCodecs];
+}
+
++ (NSArray<RTC_OBJC_TYPE(RTCVideoCodecInfo) *> *)supportedCodecsForDefaultEncoderFactory {
+    NSMutableArray<RTC_OBJC_TYPE(RTCVideoCodecInfo) *> *codecs = [[NSMutableArray alloc] init];
+
+    // H264
+    {
+        {
+            auto profile_level_id = webrtc::ParseH264ProfileLevelId([kRTCMaxSupportedH264ProfileLevelConstrainedHigh UTF8String]);
+            if (profile_level_id) {
+                auto format = webrtc::CreateH264Format(profile_level_id->profile, profile_level_id->level, "1", true);
+                RTCVideoCodecInfo *codec = [[RTCVideoCodecInfo alloc] initWithNativeSdpVideoFormat: format];
+                [codecs addObject: codec];
+            }
+        }
+        {
+            auto profile_level_id = webrtc::ParseH264ProfileLevelId([kRTCMaxSupportedH264ProfileLevelConstrainedBaseline UTF8String]);
+            if (profile_level_id) {
+                auto format = webrtc::CreateH264Format(profile_level_id->profile, profile_level_id->level, "1", true);
+                RTCVideoCodecInfo *codec = [[RTCVideoCodecInfo alloc] initWithNativeSdpVideoFormat: format];
+                [codecs addObject: codec];
+            }
+        }
+    }
+
+    // VP8
+    {
+        absl::InlinedVector<webrtc::ScalabilityMode, webrtc::kScalabilityModeCount> scalability_modes;
+        for (const auto scalability_mode : webrtc::kVP8SupportedScalabilityModes) {
+            scalability_modes.push_back(scalability_mode);
+        }
+        webrtc::SdpVideoFormat format(webrtc::SdpVideoFormat::VP8(), scalability_modes);
+        RTCVideoCodecInfo *codec = [[RTCVideoCodecInfo alloc] initWithNativeSdpVideoFormat: format];
+        [codecs addObject: codec];
+    }
+    // VP9
+    for (const webrtc::SdpVideoFormat& format : webrtc::SupportedVP9Codecs(true)) {
+        RTCVideoCodecInfo *codec = [[RTCVideoCodecInfo alloc] initWithNativeSdpVideoFormat: format];
+        [codecs addObject: codec];
+    }
+    // AV1
+    {
+        webrtc::SdpVideoFormat format(
+            webrtc::kAv1CodecName, webrtc::CodecParameterMap(),
+            webrtc::LibaomAv1EncoderSupportedScalabilityModes());
+        RTCVideoCodecInfo *codec = [[RTCVideoCodecInfo alloc] initWithNativeSdpVideoFormat: format];
+        [codecs addObject: codec];
+    }
+
+    return [codecs copy];
+}
+
+@end
