diff --git a/sdk/BUILD.gn b/sdk/BUILD.gn
index 358338f447..11714e391e 100644
--- a/sdk/BUILD.gn
+++ b/sdk/BUILD.gn
@@ -914,6 +914,8 @@ if (is_ios || is_mac) {
         "objc/api/peerconnection/RTCPeerConnection+Stats.mm",
         "objc/api/peerconnection/RTCPeerConnection.h",
         "objc/api/peerconnection/RTCPeerConnection.mm",
+        "objc/api/peerconnection/RTCPeerConnectionDependencies.h",
+        "objc/api/peerconnection/RTCPeerConnectionDependencies.mm",
         "objc/api/peerconnection/RTCPeerConnectionFactory+Native.h",
         "objc/api/peerconnection/RTCPeerConnectionFactory+Private.h",
         "objc/api/peerconnection/RTCPeerConnectionFactory.h",
@@ -956,6 +958,7 @@ if (is_ios || is_mac) {
         "objc/api/peerconnection/RTCSessionDescription+Private.h",
         "objc/api/peerconnection/RTCSessionDescription.h",
         "objc/api/peerconnection/RTCSessionDescription.mm",
+        "objc/api/peerconnection/RTCSSLCertificateVerifier.h",
         "objc/api/peerconnection/RTCStatisticsReport+Private.h",
         "objc/api/peerconnection/RTCStatisticsReport.h",
         "objc/api/peerconnection/RTCStatisticsReport.mm",
@@ -1250,6 +1253,7 @@ if (is_ios || is_mac) {
           "objc/components/video_codec/RTCVideoDecoderH264.h",
           "objc/components/video_codec/RTCVideoEncoderFactoryH264.h",
           "objc/components/video_codec/RTCVideoEncoderH264.h",
+          "objc/components/video_codec/RTCVideoEncoderFactorySimulcast.h",
           "objc/components/video_frame_buffer/RTCCVPixelBuffer.h",
           "objc/helpers/RTCCameraPreviewView.h",
           "objc/helpers/RTCDispatcher.h",
diff --git a/sdk/objc/api/peerconnection/RTCPeerConnection+Private.h b/sdk/objc/api/peerconnection/RTCPeerConnection+Private.h
index 558cf07dce..2f530b5d22 100644
--- a/sdk/objc/api/peerconnection/RTCPeerConnection+Private.h
+++ b/sdk/objc/api/peerconnection/RTCPeerConnection+Private.h
@@ -12,6 +12,8 @@
 
 #include "api/peer_connection_interface.h"
 
+@class RTC_OBJC_TYPE (RTCPeerConnectionDependencies);
+
 NS_ASSUME_NONNULL_BEGIN
 
 namespace webrtc {
@@ -83,6 +85,12 @@ class PeerConnectionDelegateAdapter : public PeerConnectionObserver {
                                 delegate:
                                     (nullable id<RTC_OBJC_TYPE(RTCPeerConnectionDelegate)>)delegate;
 
+- (nullable instancetype)initWithFactory:(RTC_OBJC_TYPE(RTCPeerConnectionFactory) *)factory
+                           configuration:(RTC_OBJC_TYPE(RTCConfiguration) *)configuration
+                             constraints:(RTC_OBJC_TYPE(RTCMediaConstraints) *)constraints
+                            dependencies:(RTC_OBJC_TYPE(RTCPeerConnectionDependencies) *)dependencies
+                                delegate:(nullable id<RTC_OBJC_TYPE(RTCPeerConnectionDelegate)>)delegate;
+
 /** Initialize an RTCPeerConnection with a configuration, constraints,
  *  delegate and PeerConnectionDependencies.
  */
diff --git a/sdk/objc/api/peerconnection/RTCPeerConnection.mm b/sdk/objc/api/peerconnection/RTCPeerConnection.mm
index 05fe581d08..ee315ecae6 100644
--- a/sdk/objc/api/peerconnection/RTCPeerConnection.mm
+++ b/sdk/objc/api/peerconnection/RTCPeerConnection.mm
@@ -328,6 +328,23 @@ - (nullable instancetype)initWithFactory:(RTC_OBJC_TYPE(RTCPeerConnectionFactory
                            delegate:delegate];
 }
 
+- (nullable instancetype)
+    initWithFactory:(RTC_OBJC_TYPE(RTCPeerConnectionFactory) *)factory
+      configuration:(RTC_OBJC_TYPE(RTCConfiguration) *)configuration
+        constraints:(RTC_OBJC_TYPE(RTCMediaConstraints) *)constraints
+       dependencies:(RTC_OBJC_TYPE(RTCPeerConnectionDependencies) *)dependencies
+           delegate:(nullable id<RTC_OBJC_TYPE(RTCPeerConnectionDelegate)>)delegate {
+  NSParameterAssert(factory);
+  std::unique_ptr<webrtc::PeerConnectionDependencies> nativeDependencies =
+    std::make_unique<webrtc::PeerConnectionDependencies>(nullptr);
+  // TODO: dependencies を native に変換する
+  return [self initWithDependencies:factory
+                      configuration:configuration
+                        constraints:constraints
+                       dependencies:std::move(nativeDependencies)
+                           delegate:delegate];
+}
+
 - (nullable instancetype)
     initWithDependencies:(RTC_OBJC_TYPE(RTCPeerConnectionFactory) *)factory
            configuration:(RTC_OBJC_TYPE(RTCConfiguration) *)configuration
diff --git a/sdk/objc/api/peerconnection/RTCPeerConnectionDependencies.h b/sdk/objc/api/peerconnection/RTCPeerConnectionDependencies.h
new file mode 100644
index 0000000000..9ec6d4ab2a
--- /dev/null
+++ b/sdk/objc/api/peerconnection/RTCPeerConnectionDependencies.h
@@ -0,0 +1,18 @@
+#import <Foundation/Foundation.h>
+
+#import "RTCMacros.h"
+
+@class RTC_OBJC_TYPE(RTCSSLCertificateVerifier);
+
+NS_ASSUME_NONNULL_BEGIN
+
+RTC_OBJC_EXPORT
+@interface RTC_OBJC_TYPE (RTCPeerConnectionDependencies) : NSObject
+
+@property (readonly, nonatomic) RTC_OBJC_TYPE(RTCSSLCertificateVerifier) *sslCertificateVerifier;
+
+- (instancetype)initWithSSLCertificateVerifier:(RTC_OBJC_TYPE(RTCSSLCertificateVerifier) *)sslCertificateVerifier;
+
+@end
+
+NS_ASSUME_NONNULL_END
diff --git a/sdk/objc/api/peerconnection/RTCPeerConnectionDependencies.mm b/sdk/objc/api/peerconnection/RTCPeerConnectionDependencies.mm
new file mode 100644
index 0000000000..fdf74819e7
--- /dev/null
+++ b/sdk/objc/api/peerconnection/RTCPeerConnectionDependencies.mm
@@ -0,0 +1,18 @@
+#import "RTCPeerConnectionDependencies.h"
+
+@interface RTC_OBJC_TYPE (RTCPeerConnectionDependencies) ()
+
+@property (nonatomic) RTC_OBJC_TYPE(RTCSSLCertificateVerifier) *sslCertificateVerifier;
+
+@end
+
+@implementation RTC_OBJC_TYPE (RTCPeerConnectionDependencies)
+
+- (instancetype)initWithSSLCertificateVerifier:(RTC_OBJC_TYPE(RTCSSLCertificateVerifier) *)sslCertificateVerifier {
+    if (self = [super init]) {
+        self.sslCertificateVerifier = sslCertificateVerifier;
+    }
+    return self;
+}
+
+@end
diff --git a/sdk/objc/api/peerconnection/RTCPeerConnectionFactory.h b/sdk/objc/api/peerconnection/RTCPeerConnectionFactory.h
index 2b0489885f..516c30e8df 100644
--- a/sdk/objc/api/peerconnection/RTCPeerConnectionFactory.h
+++ b/sdk/objc/api/peerconnection/RTCPeerConnectionFactory.h
@@ -22,6 +22,7 @@ NS_ASSUME_NONNULL_BEGIN
 @class RTC_OBJC_TYPE(RTCPeerConnection);
 @class RTC_OBJC_TYPE(RTCVideoSource);
 @class RTC_OBJC_TYPE(RTCVideoTrack);
+@class RTC_OBJC_TYPE (RTCPeerConnectionDependencies);
 @class RTC_OBJC_TYPE(RTCPeerConnectionFactoryOptions);
 @protocol RTC_OBJC_TYPE
 (RTCPeerConnectionDelegate);
@@ -75,6 +76,12 @@ RTC_OBJC_EXPORT
                         constraints:(RTC_OBJC_TYPE(RTCMediaConstraints) *)constraints
                            delegate:(nullable id<RTC_OBJC_TYPE(RTCPeerConnectionDelegate)>)delegate;
 
+- (RTC_OBJC_TYPE(RTCPeerConnection) *)
+    peerConnectionWithConfiguration:(RTC_OBJC_TYPE(RTCConfiguration) *)configuration
+                        constraints:(RTC_OBJC_TYPE(RTCMediaConstraints) *)constraints
+                       dependencies:(RTC_OBJC_TYPE(RTCPeerConnectionDependencies) *)dependencies
+                           delegate:(nullable id<RTC_OBJC_TYPE(RTCPeerConnectionDelegate)>)delegate;
+
 /** Set the options to be used for subsequently created RTCPeerConnections */
 - (void)setOptions:(nonnull RTC_OBJC_TYPE(RTCPeerConnectionFactoryOptions) *)options;

diff --git a/sdk/objc/api/peerconnection/RTCPeerConnectionFactory.mm b/sdk/objc/api/peerconnection/RTCPeerConnectionFactory.mm
index 2f324f7289..d089921223 100644
--- a/sdk/objc/api/peerconnection/RTCPeerConnectionFactory.mm
+++ b/sdk/objc/api/peerconnection/RTCPeerConnectionFactory.mm
@@ -279,6 +279,18 @@ - (instancetype)initWithNativeAudioEncoderFactory:
                                                                delegate:delegate];
 }
 
+- (RTC_OBJC_TYPE(RTCPeerConnection) *)
+    peerConnectionWithConfiguration:(RTC_OBJC_TYPE(RTCConfiguration) *)configuration
+        constraints:(RTC_OBJC_TYPE(RTCMediaConstraints) *)constraints
+        dependencies:(RTC_OBJC_TYPE(RTCPeerConnectionDependencies) *)dependencies
+        delegate:(nullable id<RTC_OBJC_TYPE(RTCPeerConnectionDelegate)>)delegate {
+    return [[RTC_OBJC_TYPE(RTCPeerConnection) alloc] initWithFactory:self
+                                                       configuration:configuration
+                                                         constraints:constraints
+                                                        dependencies:dependencies
+                                                            delegate:delegate];
+}
+
 - (void)setOptions:(nonnull RTC_OBJC_TYPE(RTCPeerConnectionFactoryOptions) *)options {
   RTC_DCHECK(options != nil);
   _nativeFactory->SetOptions(options.nativeOptions);

diff --git a/sdk/objc/api/peerconnection/RTCSSLCertificateVerifier.h b/sdk/objc/api/peerconnection/RTCSSLCertificateVerifier.h
new file mode 100644
index 0000000000..1a7047f4e3
--- /dev/null
+++ b/sdk/objc/api/peerconnection/RTCSSLCertificateVerifier.h
@@ -0,0 +1,14 @@
+#import <Foundation/Foundation.h>
+
+#import "RTCMacros.h"
+
+NS_ASSUME_NONNULL_BEGIN
+
+RTC_OBJC_EXPORT
+@protocol RTC_OBJC_TYPE (RTCSSLCertificateVerifier)
+
+- (BOOL)verifyCertificate:(NSData *)certificate;
+
+@end
+
+NS_ASSUME_NONNULL_END
