diff --git a/config/ios/BUILD.gn b/config/ios/BUILD.gn
index 6d63abf33..2d7e33b73 100644
--- a/config/ios/BUILD.gn
+++ b/config/ios/BUILD.gn
@@ -126,7 +114,7 @@ config("runtime_library") {
   # TODO(crbug.com/1223481): Temporarily use a different
   # libclang_rt.iossim.a for arm64 simulator builds. This can be
   # removed when an arm64 slice is added to upstream Clang.
-  if (target_environment == "simulator" && current_cpu == "arm64") {
+  if (!use_xcode_clang && target_environment == "simulator" && current_cpu == "arm64") {
     assert(xcode_version_int == 1300)
     ldflags += [
       "-lSystem",
