diff --git a/sdk/objc/components/video_codec/nalu_rewriter.cc b/sdk/objc/components/video_codec/nalu_rewriter.cc
index 301dd7f8c9..56442be6a7 100644
--- a/sdk/objc/components/video_codec/nalu_rewriter.cc
+++ b/sdk/objc/components/video_codec/nalu_rewriter.cc
@@ -481,7 +481,7 @@ public:
       bool bitstream_restriction_flag { false };
       uint64_t max_num_reorder_frames { 0 };
     };
-    static absl::optional<State> Parse(const std::vector<uint8_t>& unpacked_buffer)
+    static std::optional<State> Parse(const std::vector<uint8_t>& unpacked_buffer)
     {
       BitstreamReader reader(unpacked_buffer);
       auto spsState = ParseSpsUpToVui(reader);