diff --git a/build/config/clang/clang.gni b/build/config/clang/clang.gni
index 9928fc37b..0f4c72b01 100644
--- a/build/config/clang/clang.gni
+++ b/build/config/clang/clang.gni
@@ -11,9 +11,9 @@ declare_args() {
   # Indicates if the build should use the Chrome-specific plugins for enforcing
   # coding guidelines, etc. Only used when compiling with Chrome's Clang, not
   # Chrome OS's.
-  clang_use_chrome_plugins =
-      is_clang && !is_nacl && current_os != "zos" &&
-      default_toolchain != "//build/toolchain/cros:target"
+  clang_use_chrome_plugins = false
+      # is_clang && !is_nacl && current_os != "zos" &&
+      # default_toolchain != "//build/toolchain/cros:target"
 
   # Use this instead of clang_use_chrome_plugins to enable just the raw-ptr-plugin.
   clang_use_raw_ptr_plugin = false
diff --git a/build/config/compiler/BUILD.gn b/build/config/compiler/BUILD.gn
index 3b809a3c0..5c20f6d16 100644
--- a/build/config/compiler/BUILD.gn
+++ b/build/config/compiler/BUILD.gn
@@ -475,8 +475,8 @@ config("compiler") {
     # faster.
     # On arm64, this is already the default behavior.
     if (current_cpu == "x64") {
-      asmflags += [ "-femit-dwarf-unwind=no-compact-unwind" ]
-      cflags += [ "-femit-dwarf-unwind=no-compact-unwind" ]
+      #asmflags += [ "-femit-dwarf-unwind=no-compact-unwind" ]
+      #cflags += [ "-femit-dwarf-unwind=no-compact-unwind" ]
     }
 
     # dsymutil is not available in the system, on bots, for rustc to call. Our
@@ -562,55 +562,6 @@ config("compiler") {
     ldflags += [ "-Wl,-z,keep-text-section-prefix" ]
   }
 
-  if (is_clang && !is_nacl) {
-    cflags += [ "-fcrash-diagnostics-dir=" + clang_diagnostic_dir ]
-    if (save_reproducers_on_lld_crash && use_lld) {
-      ldflags += [
-        "-fcrash-diagnostics=all",
-        "-fcrash-diagnostics-dir=" + clang_diagnostic_dir,
-      ]
-    }
-
-    # TODO(hans): Remove this once Clang generates better optimized debug info
-    # by default. https://crbug.com/765793
-    cflags += [
-      "-mllvm",
-      "-instcombine-lower-dbg-declare=0",
-    ]
-    if (!is_debug && use_thin_lto && is_a_target_toolchain) {
-      if (is_win) {
-        ldflags += [ "-mllvm:-instcombine-lower-dbg-declare=0" ]
-      } else {
-        ldflags += [ "-Wl,-mllvm,-instcombine-lower-dbg-declare=0" ]
-      }
-    }
-
-    # TODO(crbug.com/40283598): This causes binary size growth and potentially
-    # other problems.
-    # TODO(crbug.com/40284925): This isn't supported by Cronet's mainline llvm version.
-    if (default_toolchain != "//build/toolchain/cros:target" &&
-        !llvm_android_mainline) {
-      cflags += [
-        "-mllvm",
-        "-split-threshold-for-reg-with-hint=0",
-      ]
-      if (use_thin_lto && is_a_target_toolchain) {
-        if (is_win) {
-          ldflags += [ "-mllvm:-split-threshold-for-reg-with-hint=0" ]
-        } else {
-          ldflags += [ "-Wl,-mllvm,-split-threshold-for-reg-with-hint=0" ]
-        }
-      }
-    }
-
-    # TODO(crbug.com/40192287): Investigate why/if this should be needed.
-    if (is_win) {
-      cflags += [ "/clang:-ffp-contract=off" ]
-    } else {
-      cflags += [ "-ffp-contract=off" ]
-    }
-  }
-
   # C11/C++11 compiler flags setup.
   # ---------------------------
   if (is_linux || is_chromeos || is_android || (is_nacl && is_clang) ||
@@ -1997,7 +1948,7 @@ config("default_warnings") {
 
   # TODO(crbug.com/355446806): Clean up and enable.
   if (is_apple) {
-    ldflags += [ "-Wl,-no_warn_duplicate_libraries" ]
+    # ldflags += [ "-Wl,-no_warn_duplicate_libraries" ]
   }
 
   # Rust warnings
diff --git a/build/toolchain/apple/toolchain.gni b/build/toolchain/apple/toolchain.gni
index ab7378fe2..9b2a869d9 100644
--- a/build/toolchain/apple/toolchain.gni
+++ b/build/toolchain/apple/toolchain.gni
@@ -153,7 +153,8 @@ template("single_apple_toolchain") {
       toolchain_is_component_build = is_component_build
     }
 
-    prefix = rebase_path("$clang_base_path/bin/", root_build_dir)
+    # prefix = rebase_path("$clang_base_path/bin/", root_build_dir)
+    prefix = invoker.bin_path
     _cc = "${prefix}clang"
     _cxx = "${prefix}clang++"
 
diff --git a/build/toolchain/mac/BUILD.gn b/build/toolchain/mac/BUILD.gn
index d8543c4f2..db6c76528 100644
--- a/build/toolchain/mac/BUILD.gn
+++ b/build/toolchain/mac/BUILD.gn
@@ -24,7 +24,7 @@ template("mac_toolchain") {
 
       if (target_os == "ios") {
         # Use LLD for the host part of a chrome/ios build.
-        use_lld = true
+        use_lld = false
 
         # Override `is_component_build` for the host toolchain.
         # See https://crbug.com/gn/286 for details why this is
diff --git a/third_party/nasm/BUILD.gn b/third_party/nasm/BUILD.gn
index 3834c64b..6f74500f 100644
--- a/third_party/nasm/BUILD.gn
+++ b/third_party/nasm/BUILD.gn
@@ -84,7 +84,7 @@ config("nasm_config") {
       "-Wno-uninitialized",
 
       # NASM sometimes set variables but doesn't use them.
-      "-Wno-unused-but-set-variable",
+      # "-Wno-unused-but-set-variable",
 
       # NASM undefines __STRICT_ANSI__
       "-Wno-builtin-macro-redefined",
