diff --git a/build/config/compiler/BUILD.gn b/build/config/compiler/BUILD.gn
index 8048ec5e8..c0bb45ac2 100644
--- a/build/config/compiler/BUILD.gn
+++ b/build/config/compiler/BUILD.gn
@@ -534,7 +534,7 @@ config("compiler") {
     ldflags += [ "-Wl,-z,keep-text-section-prefix" ]
   }
 
-  if (is_clang && !is_nacl && current_os != "zos") {
+  if (is_clang && !is_nacl && current_os != "zos" && !(is_ios && enable_ios_bitcode)) {
     cflags += [ "-fcrash-diagnostics-dir=" + clang_diagnostic_dir ]
 
     # TODO(hans): Remove this once Clang generates better optimized debug info
diff --git a/third_party/dav1d/BUILD.gn b/third_party/dav1d/BUILD.gn
index 933c8f2f566..d752194d95b 100644
--- a/third_party/dav1d/BUILD.gn
+++ b/third_party/dav1d/BUILD.gn
@@ -36,29 +36,8 @@ if (current_cpu == "x86" || current_cpu == "x64") {
     needs_stack_alignment = true
     if (current_cpu == "x86") {
       stack_alignment = 16
-
-      if (!is_clang) {
-        # Values used by GCC.
-        preferred_stack_boundary = 4
-        incoming_stack_boundary = 2
-      }
     } else if (current_cpu == "x64") {
       stack_alignment = 32
-
-      if (!is_clang) {
-        # Values used by GCC.
-        preferred_stack_boundary = 5
-        incoming_stack_boundary = 4
-      }
-    }
-
-    if (is_clang) {
-      stackalign_flag = "-mstack-alignment=$stack_alignment"
-      stackrealign_flag = "-mstackrealign"
-    } else {
-      # Assume GCC for now.
-      stackalign_flag = "-mpreferred-stack-boundary=$preferred_stack_boundary"
-      stackrealign_flag = "-mincoming-stack-boundary=$incoming_stack_boundary"
     }
   }
 } else {
@@ -112,9 +91,6 @@ if (is_win) {
   }
 } else {
   dav1d_copts += [ "-std=c99" ]
-  if (needs_stack_alignment) {
-    dav1d_copts += [ stackalign_flag ]
-  }
   if (is_mac || is_ios) {
     dav1d_copts += [ "-D_DARWIN_C_SOURCE" ]
   }
@@ -179,10 +155,6 @@ static_library("dav1d_entrypoints") {
     sources += [ "libdav1d/src/win32/thread.c" ]
   }
 
-  if (needs_stack_alignment) {
-    cflags += [ stackrealign_flag ]
-  }
-
   deps = [ ":dav1d_headers" ]
 }
