diff --git a/BUILD.gn b/BUILD.gn
index d566340..8885a52 100644
--- a/BUILD.gn
+++ b/BUILD.gn
@@ -12,6 +12,10 @@ if (current_cpu == "arm" || current_cpu == "arm64") {
 
 assert(!is_ios, "This is not used on iOS, don't drag it in unintentionally")
 
+config("libjpeg_headers_config") {
+  defines = [ "MANGLE_JPEG_NAMES" ]
+}
+
 source_set("libjpeg_headers") {
   sources = [
     "jconfig.h",
@@ -21,7 +25,7 @@ source_set("libjpeg_headers") {
     "jpeglib.h",
     "jpeglibmangler.h",
   ]
-  defines = [ "MANGLE_JPEG_NAMES" ]
+  public_configs = [ ":libjpeg_headers_config" ]
 }
 
 if (current_cpu == "x86" || current_cpu == "x64") {