diff --git a/src/sdk/objc/Framework/Headers/WebRTC/RTCVideoDecoderAV1.h b/src/sdk/objc/Framework/Headers/WebRTC/RTCVideoDecoderAV1.h
new file mode 100644
index 0000000000..6454724dc5
--- /dev/null
+++ b/src/sdk/objc/Framework/Headers/WebRTC/RTCVideoDecoderAV1.h
@@ -0,0 +1,11 @@
+/*
+ *  Copyright 2017 The WebRTC project authors. All Rights Reserved.
+ *
+ *  Use of this source code is governed by a BSD-style license
+ *  that can be found in the LICENSE file in the root of the source
+ *  tree. An additional intellectual property rights grant can be found
+ *  in the file PATENTS.  All contributing project authors may
+ *  be found in the AUTHORS file in the root of the source tree.
+ */
+
+#import "api/video_codec/RTCVideoDecoderAV1.h"
diff --git a/src/sdk/objc/Framework/Headers/WebRTC/RTCVideoEncoderAV1.h b/src/sdk/objc/Framework/Headers/WebRTC/RTCVideoEncoderAV1.h
new file mode 100644
index 0000000000..2059bf763b
--- /dev/null
+++ b/src/sdk/objc/Framework/Headers/WebRTC/RTCVideoEncoderAV1.h
@@ -0,0 +1,11 @@
+/*
+ *  Copyright 2017 The WebRTC project authors. All Rights Reserved.
+ *
+ *  Use of this source code is governed by a BSD-style license
+ *  that can be found in the LICENSE file in the root of the source
+ *  tree. An additional intellectual property rights grant can be found
+ *  in the file PATENTS.  All contributing project authors may
+ *  be found in the AUTHORS file in the root of the source tree.
+ */
+
+#import "api/video_codec/RTCVideoEncoderAV1.h"
diff --git a/src/sdk/objc/api/peerconnection/RTCRtpCodecParameters.h b/src/sdk/objc/api/peerconnection/RTCRtpCodecParameters.h
index 5d3cac5c96..379ae1e1f2 100644
--- a/src/sdk/objc/api/peerconnection/RTCRtpCodecParameters.h
+++ b/src/sdk/objc/api/peerconnection/RTCRtpCodecParameters.h
@@ -29,6 +29,7 @@ RTC_EXTERN const NSString *const kRTCDtmfCodecName;
 RTC_EXTERN const NSString *const kRTCComfortNoiseCodecName;
 RTC_EXTERN const NSString *const kRTCVp8CodecName;
 RTC_EXTERN const NSString *const kRTCVp9CodecName;
+RTC_EXTERN const NSString *const kRTCAv1CodecName;
 RTC_EXTERN const NSString *const kRTCH264CodecName;
 
 /** Defined in http://w3c.github.io/webrtc-pc/#idl-def-RTCRtpCodecParameters */
diff --git a/src/sdk/objc/api/peerconnection/RTCRtpCodecParameters.mm b/src/sdk/objc/api/peerconnection/RTCRtpCodecParameters.mm
index f25679e329..900f6c29cc 100644
--- a/src/sdk/objc/api/peerconnection/RTCRtpCodecParameters.mm
+++ b/src/sdk/objc/api/peerconnection/RTCRtpCodecParameters.mm
@@ -32,6 +32,7 @@
     @(cricket::kComfortNoiseCodecName);
 const NSString * const kRTCVp8CodecName = @(cricket::kVp8CodecName);
 const NSString * const kRTCVp9CodecName = @(cricket::kVp9CodecName);
+const NSString * const kRTCAv1CodecName = @(cricket::kAv1CodecName);
 const NSString * const kRTCH264CodecName = @(cricket::kH264CodecName);
 
 @implementation RTCRtpCodecParameters
