diff --git a/config/compiler/BUILD.gn b/config/compiler/BUILD.gn
index 944a039d0..ecf87d683 100644
--- a/config/compiler/BUILD.gn
+++ b/config/compiler/BUILD.gn
@@ -426,8 +426,8 @@ config("compiler") {
     # faster.
     # On arm64, this is already the default behavior.
     if (current_cpu == "x64") {
-      asmflags += [ "-femit-dwarf-unwind=no-compact-unwind" ]
-      cflags += [ "-femit-dwarf-unwind=no-compact-unwind" ]
+      #asmflags += [ "-femit-dwarf-unwind=no-compact-unwind" ]
+      #cflags += [ "-femit-dwarf-unwind=no-compact-unwind" ]
     }
   }
 
@@ -529,37 +529,6 @@ config("compiler") {
     ldflags += [ "-Wl,-z,keep-text-section-prefix" ]
   }
 
-  if (is_clang && !is_nacl && current_os != "zos") {
-    cflags += [ "-fcrash-diagnostics-dir=" + clang_diagnostic_dir ]
-    if (save_reproducers_on_lld_crash && use_lld) {
-      ldflags += [
-        "-fcrash-diagnostics=all",
-        "-fcrash-diagnostics-dir=" + clang_diagnostic_dir,
-      ]
-    }
-
-    # TODO(hans): Remove this once Clang generates better optimized debug info
-    # by default. https://crbug.com/765793
-    cflags += [
-      "-mllvm",
-      "-instcombine-lower-dbg-declare=0",
-    ]
-    if (!is_debug && use_thin_lto && is_a_target_toolchain) {
-      if (is_win) {
-        ldflags += [ "-mllvm:-instcombine-lower-dbg-declare=0" ]
-      } else {
-        ldflags += [ "-Wl,-mllvm,-instcombine-lower-dbg-declare=0" ]
-      }
-    }
-
-    # TODO(crbug.com/1235145): Investigate why/if this should be needed.
-    if (is_win) {
-      cflags += [ "/clang:-ffp-contract=off" ]
-    } else {
-      cflags += [ "-ffp-contract=off" ]
-    }
-  }
-
   # Rust compiler setup (for either clang or rustc).
   if (enable_rust) {
     defines += [ "RUST_ENABLED" ] 

diff --git a/toolchain/apple/toolchain.gni b/toolchain/apple/toolchain.gni
index 399a62f03..fcb2fdf2c 100644
--- a/toolchain/apple/toolchain.gni
+++ b/toolchain/apple/toolchain.gni
@@ -121,7 +121,8 @@ template("apple_toolchain") {
       toolchain_is_component_build = is_component_build
     }
 
-    prefix = rebase_path("$clang_base_path/bin/", root_build_dir)
+    #prefix = rebase_path("$clang_base_path/bin/", root_build_dir)
+    prefix = invoker.bin_path
     _cc = "${prefix}clang"
     _cxx = "${prefix}clang++"
 
