diff --git a/config/compiler/BUILD.gn b/config/compiler/BUILD.gn
index f378b9560..cfe925698 100644
--- a/config/compiler/BUILD.gn
+++ b/config/compiler/BUILD.gn
@@ -1531,7 +1531,7 @@ config("default_warnings") {
 
       if (!is_nacl) {
         # TODO(https://crbug.com/1300731) Clean up and enable.
-        cflags += [ "-Wno-unqualified-std-cast-call" ]
+        #cflags += [ "-Wno-unqualified-std-cast-call" ]
       }
 
       if (!is_nacl && !(is_chromeos ||
@@ -1539,7 +1539,7 @@ config("default_warnings") {
         # TODO(https://crbug.com/1316298): Re-enable once test failure is figured out
         cflags += [
           "-Xclang",
-          "-no-opaque-pointers",
+          #"-no-opaque-pointers",
         ]
       }
 
diff --git a/toolchain/apple/toolchain.gni b/toolchain/apple/toolchain.gni
index 399a62f03..fcb2fdf2c 100644
--- a/toolchain/apple/toolchain.gni
+++ b/toolchain/apple/toolchain.gni
@@ -121,7 +121,8 @@ template("apple_toolchain") {
       toolchain_is_component_build = is_component_build
     }
 
-    prefix = rebase_path("$clang_base_path/bin/", root_build_dir)
+    #prefix = rebase_path("$clang_base_path/bin/", root_build_dir)
+    prefix = invoker.bin_path
     _cc = "${prefix}clang"
     _cxx = "${prefix}clang++"
 
