diff --git a/build/toolchain/win/toolchain.gni b/build/toolchain/win/toolchain.gni
index b6874dc6b..7e9e7993a 100644
--- a/build/toolchain/win/toolchain.gni
+++ b/build/toolchain/win/toolchain.gni
@@ -194,7 +194,7 @@ template("msvc_toolchain") {
       # clang header updates happen via changing command line flags.
       show_includes = "/showIncludes:user"
     } else {
-      show_includes = "/showIncludes"
+      show_includes = ""
     }
 
     tool("cc") {
diff --git a/BUILD.gn b/BUILD.gn
index 9c55478c25..6512ee25ea 100644
--- a/BUILD.gn
+++ b/BUILD.gn
@@ -517,7 +517,6 @@ if (!build_with_chromium) {
       "pc:rtc_pc",
       "sdk",
       "video",
-      "modules/audio_device:audio_device_module_from_input_and_output",
       "//third_party/zlib",
       "rtc_base:log_sinks",
       "media:rtc_simulcast_encoder_adapter",
diff --git a/modules/audio_device/BUILD.gn b/modules/audio_device/BUILD.gn
index d441479b33..2925586b0e 100644
--- a/modules/audio_device/BUILD.gn
+++ b/modules/audio_device/BUILD.gn
@@ -306,10 +306,17 @@ rtc_library("audio_device_impl") {
         ]
       }
       if (is_win) {
-        sources += [
-          "win/audio_device_core_win.cc",
-          "win/audio_device_core_win.h",
-        ]
+        if (target_os == "winuwp") {
+          sources += [
+            "winuwp/audio_device_core_win.cc",
+            "winuwp/audio_device_core_win.h",
+          ]
+        } else {
+          sources += [
+            "win/audio_device_core_win.cc",
+            "win/audio_device_core_win.h",
+          ]
+        }
         libs = [
           # Required for the built-in WASAPI AEC.
           "dmoguids.lib",
diff --git a/modules/audio_device/audio_device_impl.cc b/modules/audio_device/audio_device_impl.cc
index 80ed928933..50bf744032 100644
--- a/modules/audio_device/audio_device_impl.cc
+++ b/modules/audio_device/audio_device_impl.cc
@@ -22,8 +22,12 @@
 
 #if defined(_WIN32)
 #if defined(WEBRTC_WINDOWS_CORE_AUDIO_BUILD)
+#if defined(WINUWP)
+#include "modules/audio_device/winuwp/audio_device_core_win.h"
+#else
 #include "modules/audio_device/win/audio_device_core_win.h"
 #endif
+#endif
 #elif defined(WEBRTC_ANDROID)
 #include <stdlib.h>
 #include "sdk/android/native_api/audio_device_module/audio_device_android.h"
diff --git a/build/config/win/BUILD.gn b/build/config/win/BUILD.gn
index 1e76a54cc..b254b3ac8 100644
--- a/build/config/win/BUILD.gn
+++ b/build/config/win/BUILD.gn
@@ -245,7 +245,11 @@ config("runtime_library") {
     } else {
       defines += [ "WINAPI_FAMILY=WINAPI_FAMILY_DESKTOP_APP" ]
     }
-    cflags_cc += [ "/EHsc" ]
+    cflags_cc += [
+      "/std:c++20",  # Required for having a wrapper using the WinRT/C++ projection.
+      "/await",
+      "/EHsc",  # Required for having a wrapper using the WinRT/C++ projection.
+    ]
 
     # This warning is given because the linker cannot tell the difference
     # between consuming WinRT APIs versus authoring WinRT within static
